experimental/FTheoryTools/src/G4Fluxes/special_attributes.jl

```diff
@@ -262,3 +101,467 @@ function ambient_space_models_of_g4_fluxes(m::AbstractFTheoryModel; check::Bool
+  list_of_base_divisor_pairs_to_be_considered = Oscar._ambient_space_base_divisor_pairs_to_be_considered(m)
+  list_of_divisor_pairs_to_be_considered = Oscar._ambient_space_divisor_pairs_to_be_considered(m)
+
+
+  # (5) Work out the relevant intersection numbers to tell if a flux is vertical
+  vertical_constraint_matrix = Vector{Vector{QQFieldElem}}()
+  if arxiv_doi(m) == "10.48550/arXiv.1511.03209"
+
+    # Use special intersection theory for special F-theory model. This technology could be extended beyond this one use-case in the future.
+    for i in 1:length(ambient_space_flux_candidates_basis)
+
+      condition = Vector{ZZRingElem}()
+      idxs = findall(x -> x != 0, collect(exponents(polynomial(ambient_space_flux_candidates_basis[i]).f))[1])
+      if length(idxs) == 1
+        idxs = [idxs[1], idxs[1]]
+      end
+
+      # Compute against pairs of base divisors
+      for j in 1:length(list_of_base_divisor_pairs_to_be_considered)
+
+        my_tuple = Tuple(sort([idxs..., list_of_base_divisor_pairs_to_be_considered[j]...]))
+        push!(condition, sophisticated_intersection_product(ambient_space(m), my_tuple, hypersurface_equation(m), inter_dict, s_inter_dict, data))
+
+      end
+
+      # Compute against zero section and base divisor
+      pos_zero_section = findfirst(x -> x == "z", string.(gS))
+      for j in 1:n_rays(base_space(m))
+
+        my_tuple = Tuple(sort([idxs..., [j, pos_zero_section]...]))
+        push!(condition, sophisticated_intersection_product(ambient_space(m), my_tuple, hypersurface_equation(m), inter_dict, s_inter_dict, data))
+
+      end
+
+      push!(vertical_constraint_matrix, condition)
+
+    end
+
+  else
+  
+    # Cover all other case with generic, but potentially painfully slow methodology.
+    tds = torusinvariant_prime_divisors(ambient_space(m))
+    cds = [cohomology_class(tds[i]) for i in 1:length(tds)]
+    pt_class = cohomology_class(anticanonical_divisor_class(ambient_space(m)))
+    for i in 1:length(ambient_space_flux_candidates_basis)
+
+      condition = Vector{QQFieldElem}()
+
+      idxs = findall(x -> x != 0, collect(exponents(polynomial(ambient_space_flux_candidates_basis[i]).f))[1])
+      if length(idxs) == 1
+        idxs = [idxs[1], idxs[1]]
+      end
+
+      # Compute against pairs of base divisors
+      for j in 1:length(list_of_divisor_pairs_to_be_considered)
+
+        my_tuple = Tuple(sort([idxs..., list_of_divisor_pairs_to_be_considered[j]...]))
+        if !haskey(inter_dict, my_tuple)
+          class = ambient_space_flux_candidates_basis[i] * cds[list_of_divisor_pairs_to_be_considered[j][1]] * cds[list_of_divisor_pairs_to_be_considered[j][2]] * pt_class
+          inter_dict[my_tuple] = integrate(class)
+        end
+        push!(condition, inter_dict[my_tuple])
+        
+      end
+
+      # Compute against zero section and base divisor
+      zsc = zero_section_class(m)
+      pos_zero_section = findfirst(x -> x == string(polynomial(zsc).f), string.(gS))
```

Comment:
This can fail to find the index of the zero_section, even if it has been correctly specified

Reply:
I am thinking about "vanilla scenarios": 5d toric ambient space `X_Sigma`, resolved 4d hypersurface `Y` therein and zero section given as the restriction of `V(c) subseteq X_Sigma` to `Y`, where `c` is a suitable homogeneous coordinate of the toric ambient space.

When you say "can fail", are you thinking of more general setups? E.g. setups in which the zero section is cut-out by a more complicated polynomial? Or what other types of failure are you thinking about?

And maybe most important: Suggestion for improving?

Reply:
Well, I've been testing this for the F11 toric hypersurface(hypersurface form), after specifying the zero_section_class "v" in the JSON file

Reply:
However, if the index of the zero section can be found, I have already tested the well_quantized_ambient_space_models_of_g4_fluxes for this model by taking a Z-linear combination of the output, where all coefficients are 1, and tested that with the other built-in method "passes_elementary_quantization_checks(g4)", which returns true

Reply:
I have now tried to test the well_quantized_and_vertical_ambient_space_models_of_g4_fluxes method for an example that returns something non-empty(I have chosen the QSM model with k=2021 as this one is simpler and other QSM models were too slow), but taking some Z-linear combination of the output and applying "passes_verticality_checks" returns false

Reply:
> I am thinking about "vanilla scenarios": 5d toric ambient space `X_Sigma`, resolved 4d hypersurface `Y` therein and zero section given as the restriction of `V(c) subseteq X_Sigma` to `Y`, where `c` is a suitable homogeneous coordinate of the toric ambient space.
> 
> When you say "can fail", are you thinking of more general setups? E.g. setups in which the zero section is cut-out by a more complicated polynomial? Or what other types of failure are you thinking about?
> 
> And most important: Suggestion for improving @emikelsons ?

Sorry - somehow part of your earlier messages was not displayed (to me) when writing this message. This was written as response after only seeing your first message "This can fail to find the index of the zero_section, even if it has been correctly specified", and is thus only partly relevant.

Reply:
> `pos_zero_section = findfirst(x -> x == string(polynomial(zsc)), string.([polynomial(x) for x in cds]))`

Ok. I have a feeling what you are hinting at: `polynomial(zsc)` is an element of a quotient ring. The string of this object may depend on its representation, and you are suggesting to compare with the strings of the (polynomials of) the cohomology classes in `cds`, which are also elements of the same quotient ring. Hence, they both should be presented in the same way.

In contrast, I lifted the quotient ring element (with the `.f`) to a multivariate polynomial ring and compared with the generators of the Cox ring. I believe what you are hinting at is that after `.f`, the comparison may fail, because a different representation is chosen? Is that what is happening for you?

Either way - it feels like your suggestion may be the better way to go. Of course, we could also execute a comparison test in the quotient ring. But then, we know that this will not complete for the big model. So not point in doing this.

An alternative would be to introduce an attribute `pos_of_zero_section` or alike, that we specify for all models upon construction. But I did not want to go that heavy route. At least not just yet.

Reply:
You're right in the future we should have this attribute as we already have this information in the JSON file, but we lose it during construction

Reply:
Maybe we can merge this one first, and then tend to this? Or we fix it in this PR.

Either way - I just pushed a code change with your suggested line to find the position of the zero section in the generic case.